<img src="https://idp.bits-pilani.ac.in/idp/login_files/logo.gif">


<h1><center> ML Assignment Number 2-Q1_Group 227</center></h1>

### Objective and the Data 

In this assignment we are working with a dummy advertising data set, indicating whether or not a particular internet user clicked on an Advertisement on a company website. We have created a model that will predict whether or not they will click on an ad based on the features of that user. The accuracy of the model is calculated using the test data.

This data set contains the following features:

        Daily Time Spent on Site : consumer time on site in minutes

        Age : customer age in years

        Area Income': Avg. Income of geographical area of consumer

        Daily Internet Usage': Avg. minutes a day consumer is on the internet

        Ad Topic Line': Headline of the advertisement

        City: City of consumer

        Male: Whether or not consumer was male

        Country : Country of consumer

        Timestamp : Time at which consumer clicked on Ad or closed window
        
And the following one is the target :

        Clicked on Ad : 0 or 1 indicated clicking on Ad


## Importing the Libraries



In [1]:
import numpy as np
import pandas as pd
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn import preprocessing

## Loading the Dataset and printing a sample data


In [2]:
#dataset import
print("\n Loading the dataset")
dataset = pd.read_csv('advertising-1.csv', skiprows=1, delimiter=',')
dataset.columns = ['Daily Time Spent on Site', 'Age', 'Area Income', 'Daily Internet Usage', 'Ad Topic Line', 'City', 
                   'Male', 'Country', 'Timestamp', 'Clicked on Ad']

print("\n Printing the first few lines of the dataset")
print(dataset.head())


 Loading the dataset

 Printing the first few lines of the dataset
   Daily Time Spent on Site  Age  Area Income  Daily Internet Usage  \
0                     80.23   31     68441.85                193.77   
1                     69.47   26     59785.94                236.50   
2                     74.15   29     54806.18                245.89   
3                     68.37   35     73889.99                225.58   
4                     59.99   23     59761.56                226.74   

                           Ad Topic Line            City  Male     Country  \
0     Monitored national standardization       West Jodi     1       Nauru   
1       Organic bottom-line service-desk        Davidton     0  San Marino   
2  Triple-buffered reciprocal time-frame  West Terrifurt     1       Italy   
3          Robust logistical utilization    South Manuel     0     Iceland   
4        Sharable client-driven software       Jamieberg     1      Norway   

             Timestamp  Clicked on A

## Splitting features and labels

In [3]:
# Encoding and segregating the feature and target columns
le = preprocessing.LabelEncoder()
Ad_Topic_Line_Cat = le.fit_transform(dataset['Ad Topic Line'])
City_Cat = le.fit_transform(dataset['City'])
Country_Cat = le.fit_transform(dataset['Country'])
Timestamp_Cat = le.fit_transform(dataset['Timestamp'])

#Creating a backup dataframe
dataset_backup = dataset.copy()

dataset['Ad Topic Line_Cat'] = Ad_Topic_Line_Cat
dataset['City_Cat'] = City_Cat
dataset['Country_Cat'] = Country_Cat
dataset['Timestamp_Cat'] = Timestamp_Cat

dataset = dataset.reindex(['Daily Time Spent on Site', 'Age', 'Area Income', 'Daily Internet Usage', 'Ad Topic Line_Cat', 
                          'City_Cat', 'Male', 'Country_Cat', 'Timestamp_Cat', 'Clicked on Ad'],axis=1)


feature_cols = ['Daily Time Spent on Site', 'Age', 'Area Income', 'Daily Internet Usage', 'Ad Topic Line_Cat', 
                          'City_Cat', 'Male', 'Country_Cat', 'Timestamp_Cat']

X = dataset[feature_cols]
y = dataset['Clicked on Ad']

print("\n Normalizing the feature data")
sc = preprocessing.StandardScaler()
X = sc.fit_transform(X)



 Normalizing the feature data


## Splitting Train and Test Data

In [4]:
#train - test split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.3)

## Implementing ANN Classifier using Keras

In [5]:
# Neural network
model = Sequential()
model.add(Dense(6, input_dim=9, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

history = model.fit(X_train, y_train,validation_data = (X_test,y_test), epochs=100)

y_pred = model.predict(X_test)

Epoch 1/100
22/22 [==============================] - 2s 37ms/step - loss: 0.6864 - accuracy: 0.5024 - val_loss: 0.6697 - val_accuracy: 0.5333
Epoch 2/100
22/22 [==============================] - 0s 5ms/step - loss: 0.6612 - accuracy: 0.5831 - val_loss: 0.6396 - val_accuracy: 0.5767
Epoch 3/100
22/22 [==============================] - 0s 4ms/step - loss: 0.6248 - accuracy: 0.6178 - val_loss: 0.6075 - val_accuracy: 0.6100
Epoch 4/100
22/22 [==============================] - 0s 4ms/step - loss: 0.5951 - accuracy: 0.6794 - val_loss: 0.5749 - val_accuracy: 0.6667
Epoch 5/100
22/22 [==============================] - 0s 5ms/step - loss: 0.5712 - accuracy: 0.6930 - val_loss: 0.5442 - val_accuracy: 0.7467
Epoch 6/100
22/22 [==============================] - 0s 4ms/step - loss: 0.5460 - accuracy: 0.7602 - val_loss: 0.5145 - val_accuracy: 0.8000
Epoch 7/100
22/22 [==============================] - 0s 8ms/step - loss: 0.4905 - accuracy: 0.8179 - val_loss: 0.4867 - val_accuracy: 0.8267
Epoch 8/100


## Justification of the model

We have created a Sequential Model and add layers one at a time. In this assignment we have used a fully-connected network structure with three layers.Fully connected layers are defined using the Dense class. We specify the number of neurons or nodes in the layer as the first argument, and specify the activation function using the activation argument. The first layer must have the same number of input parameter as the number of features.

We use the rectified linear unit activation function referred to as ReLU on the first two layers and the Sigmoid function in the output layer. It used to be the case that Sigmoid and Tanh activation functions were preferred for all layers. But Better performance is achieved using the ReLU activation function. We use a sigmoid on the output layer to ensure our network output is between 0 and 1 and easy to map to either a probability of class 1 or snap to a hard classification of either class with a default threshold of 0.5.

The model expects rows of data with 9 variables (the input_dim=9 argument)
The first hidden layer has 6 nodes and uses the relu activation function.
The second hidden layer has 4 nodes and uses the relu activation function.
The output layer has one node and uses the sigmoid activation function.


## Calculating the Accuracy

In [6]:
_,accuracy = model.evaluate(X_test, y_test)
print('Accuracy: %.2f' % (accuracy*100))

10/10 [==============================] - 0s 3ms/step - loss: 0.1457 - accuracy: 0.9767
Accuracy: 97.67


### We can see that the accuracy of the model is 97.67 % which indicates that this is robust model that can be used for  prediction